### PR TITLE
BL-12288 Implement authentication and core functions

### PR DESCRIPTION
**Pull request by JoshuaLicense**
| Q                | A
| ---------------- | ---
| Type             | Feature
| Ticket           | BL-12288

**Description of the change:**

Implements the `authenticate` and `register` functions.

### CODE REVIEW

File: src/Client.php [hunk 1]
```diff
@@ -3,63 +3,270 @@
...
+use Symfony\Component\Ldap\Entry;
+use Symfony\Component\Ldap\Exception\ExceptionInterface;
+use Symfony\Component\Ldap\LdapInterface;
 
 class Client implements OAuthClientInterface
 {
+    /**
+     * @var int
+     */
+    public static $tokenExpiry = 86400;
```
Comment (Hodgy) [inline src/Client.php]: Is this and the one below used at all? Or am i just being blind?

File: src/Client.php [hunk 2]
```diff
@@ -3,63 +3,270 @@
...
+    public function setTokenFactory(TokenFactoryInterface $factory): self
+    {
+        $this->tokenFactory = $factory;
+
+        return $this;
+    }
+
+    public function getTokenFactory(): TokenFactoryInterface
+    {
+        if (isset($this->tokenFactory)) {
```
Comment (Hodgy) [inline src/Client.php:207]: Any reason we aren't passing this in on the constructor?

**Review by Hodgy — CHANGES_REQUESTED** [review]

File: src/Client.php (hunk 2 above)
Comment (JoshuaLicense) [reply src/Client.php:207]: No reason, either way would be fine, I'm just a fan of doing it like this for truly optional (and defaulted) parameters. Makes it easier to change during runtime too (generally, but also functional testing).

File: src/Client.php (hunk 1 above)
Comment (JoshuaLicense) [reply src/Client.php]: No no, my bad, good spot, I'll get it removed.

**Review by Hodgy — APPROVED** [review]

**Pull request merged**